lib/rubocop/cop/style/redundant_regexp_escape.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks for redundant escapes inside Regexp literals.
+      #
+      # @example
+      #   # bad
+      #   %r{foo\/bar}
+      #
+      #   # good
+      #   %r{foo/bar}
+      #
+      #   # good
+      #   /foo\/bar/
+      #
+      #   # good
+      #   %r/foo\/bar/
+      #
+      #   # bad
+      #   /a\-b/
+      #
+      #   # good
+      #   /a-b/
+      #
+      #   # bad
+      #   /[\+\-]\d/
+      #
+      #   # good
+      #   /[+\-]\d/
+      class RedundantRegexpEscape < Cop
+        include RangeHelp
+
+        MSG_REDUNDANT_ESCAPE = 'Redundant escape inside regexp literal'
+
+        ALLOWED_ALWAYS_ESCAPES = ' []^\\#'.chars.freeze
+        ALLOWED_WITHIN_CHAR_CLASS_METACHAR_ESCAPES = '-'.chars.freeze
```

Comment:
You can always put a dash at the beginning or end of a character class and not have to escape it, right? It would require more logic in the cop, though.

Reply:
Yes, that's true - it's similar to how `^` only needs to be escaped if it is the first character of a character class. 

My two reasons for not flagging/removing such escapes is that it would involve more logic as you say, and make the resulting character classes slightly more "fragile" to changes over time - either by adding another character or by reordering the elements of the character class, e.g.:
```
[a-] => [a-b]
[-a] => [b-a]
[ab-] => [a-b]
[a^] => [^a]
```
would all change the meaning of the character class. The refactorings may not be common, but I think I fall down on allowing escapes of `^` and `-` even though they are technically redundant. What do you think? I can take a look at the implementation would be if you think it's worth flagging such cases.